libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1047,15 +1045,31 @@ string YulUtilFunctions::storageArrayPushZeroFunction(ArrayType const& _type)
 				if iszero(lt(oldLen, <maxArrayLength>)) { <panic>() }
 				sstore(array, add(oldLen, 1))
 				slot, offset := <indexAccess>(array, oldLen)
-				<storeValue>(slot, offset, <zeroValueFunction>())
+				<?isValueType>
+					<storeValue>(slot, offset, <zeroValueFunction>())
+				<!isValueType>
+					if gt(offset, 0) { invalid() }
```

Comment:
Shouldn't this rather be a compile-time check?

Reply:
Hm - this is the same question as https://github.com/ethereum/solidity/pull/9843#discussion_r499419183 .
I'm not sure how well this could be a compile-time-only check, since it comes from the ``<indexAccess>``... and treating all cases in which the offset is zero specially everywhere seems rather involved...

It would be really nice if the optimiser could remove those checks, but I'm not sure if it'll generally inline the ``<indexAccess>`` call and I don't think it'll currently realize that it'll always assign ``offset`` to zero.

What about keeping all of these as runtime checks and trying to tweak the optimiser to handle this nicely? I haven't been following the recent work of @hrkrshnn on similar optimiser steps, so maybe that's already on our radar, but: We probably should have an optimiser step that realizes that a function always assigns a "constant" to return value (on all non-reverting code paths) and then removes it from the function and instead assigns that value directly at all call sites? Actually it doesn't have to be a constant - anything "cheap" that doesn't depend on the function argument qualifies. Actually, it might even make sense to generally try to split functions with return values that don't interdepend into several functions - that might be the better alternative to "outlining", right?

Reply:
If we go for https://github.com/ethereum/solidity/pull/9914#discussion_r499463421 and it works, then this is mute for the case here, though - but there's still https://github.com/ethereum/solidity/pull/9843#discussion_r499419183